### PR TITLE
Rewrite as a full-fledged shell command

### PR DESCRIPTION
**Pull request by wvengen**
In figuring out how give back my changes while still serving the main use-case, I ended up making a more general shell tool. Having the separate steps easily accessible by command-line options made my life a lot easier when experimenting.

I hope this is a welcome contribution.
Having the different command-line options makes the presence of a configuration file superfluous. To avoid existing users having a configuration file feeling lost, I've added a README section on storing the user's full command in a shellscript.

If anyone ever might get the idea to install the script system-wide or package it, the script should work when the tre is installed e.g. in `/usr/share/lacells-creator/`, with binaries symlinked to `/usr/bin`. It always works data in the directory the user is in. Files are always referenced using one of the variables in `lib/config` (or `PREFIX`).

### CODE REVIEW

**Comment by n76** [comment]
Seems like a reasonable way to convert the earlier ad hoc scripts into a general purpose set of utilities. Thanks!

**Comment by wvengen** [comment]
Thanks for merging! I'll keep an eye on it to see if there are any issues (did some testing but perhaps I missed something). Feel free to mention me.

**Comment by IzzySoft** [comment]
Wow, good job, @wvengen !

@n76 I hope you noticed that merge reverted some of my safety-additions from the `.gitignore` file (unintentionally, I guess). May I suggest to redo that part (reflecting the changes)? Otherwise we might end up with either a big "blob" (`.csv`) or someones private data (`gen_lacells_conf`) in the repo, which would be a bit unfortunate.

For the very same reason I'd suggest to "ship" `lib/config` as `lib/config.sample`, while optionally having the "default settings" as fallback directly in the main script (in case somebody "forgets" to create a real config first). I don't like the idea providing "passwords" (or in our case API_KEYS) at command-line, as that way they enter shell history (and might get spread through too many other files where one can "forget" having put them).

**Comment by wvengen** [comment]
@IzzySoft thanks!

I might have missed some recent addition, though I tried to include all relevant changes, thanks for checking.

The idea for shipping `lib/config` as it is, is that there are defaults that work out of the box. It is not meant to be set by the user, values can be changed by the environment and/or command-line options. This can be put either in a users `~/.profile` or `~/.bashrc`, or set in a wrapper script (as suggested in the new README).

We could suggest a default name for the wrapper script, and add that to the `.gitignore` as a precaution?

**Comment by IzzySoft** [comment]
Thanks for your quick reaction, @wvengen ! Let's check for some options first. Better thinking it through before starting anything we need to revert later for some other reason :innocent:

So what about these:

1. having `lib/config` check for a user-specific config file:
    * `lib/config.user`
    * `lib/config.local`
1. having `lib/config` glob-checking for `lib/config.*` and source them all (probably not that good an idea, see below, but just mentioning the possibility)
1. introducing an additional command-line switch `-C | --config` to tell the script to source that config after having sourced our default config (maybe it doesn't even need a "switch-name", but the first/all parameter(s) not being prefixed by an "-option" is/are dealt with as config(s) to source – e.g. `./lacells-creator -d -i -c 262 -m conf/mydevice.conf [conf/specific.conf]`)
1. introducing an additional `conf/` folder for user config files. One might e.g. need to compile different databases for different devices/purposes, and thus need different configurations
1. Depending on our decision on the above, update `.gitignore` accordingly

* Option 1 will probably be sufficient for most users. Due to fixed names, it's also easy to have those excluded with `.gitignore`
* Option 2 would be more flexible for power-users, but IMHO a bit too complex (things get complicated considering e.g. backup files like `config.bak` to not be used), though the `.gitignore` part still would be easy (`lib/config.*`)

My preference would be a combination of options 1, 3, and 4. Thus there could be a "default user config file" (to be used if the new command-line switch was not specified – or generally if it exists), plus a bunch of "special ones" to be used on-demand only (via the command-line switch; either run *instead* of the default one, or *after* it). That specific directory can be put into `.gitignore` without being explicitly created, and could be mentioned in the "documentation".

I'm aware of the fact that `.gitignore` is of no relevance for most users. I just see that as a safe-guard for those participating in development, to not accidentally add what should not be added: especially "confidential information" (such as API keys), but also "unnecessary files" like downloaded `.csv`, created database files, and "archived downloads" (`.tar.gz`). It also helps keeping the output of `git status` clean. So whatever option(s) we decide on, we should make sure to include those "unnecessary files" at least.

While a wrapper-file could be used anyhow, a simple config is easier to deal with especially by "novice users" (those not that familiar with shell programming). Introducing user config file(s) we offer both ways, and users can pick the best of two worlds.

What do you think? Would be nice to also have the opinion of @n76 here – after all, it's his project :stuck_out_tongue_winking_eye:

**Comment by wvengen** [comment]
:+1: Thanks for your thorough response!

I can see that a configuration file would be useful. I like the simple `config.user` or `config.local` idea, and would even vote for putting them in the root dir.

Usually configuration is placed in `~/.config`, as specified by the [XDG Basedir](http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html) spec (+env override). We might read `~/.config/lacells/creator.conf` or something like that, but at this moment I don't think it would be used that much.

Regarding 3 - technically it seems a bit superfluous to me, sourcing a file by command-line. But I can see it being useful for users, it feels natural. I'd go for `-C | --config`.

What about:
1. source `lib/config` (program default)
2. source XDG-compliant config from `~/.config/lacells/creator.config` (user default)
3. source `<source directory>/config.local` (user default when working from the source tree)
4. source any configuration file specified on the command-line using `-C | --config`

(A next step after that might be even to introduce device profiles, e.g. a subdirectory with a configfile per device. One lacells run could generate databases for all, and upload them one by one, detecting which device is currently plugged in by device-id. But this is way beyond my usage. Also, I tend to think that using smaller pre-generated databases that can be [downloaded](https://github.com/wvengen/lacells) from mobile devices without API keys is the way to go.)

**Comment by IzzySoft** [comment]
I happily go with your steps 1-4 – looks perfect to me! The last paragraph I already have covered with my own tool (take a look at [Adebar](https://github.com/IzzySoft/Adebar) for details). DeviceID has a handicap on some devices, though: I have 3 devices which had the same ID (originally; I've managed to work-around that using a `build.prop` hack) – so if you want to implement that, you might need to explicitly exclude `0123456789ABCDEF` (which is what I often saw).

Oops, open question in step 4: as you want to use an "option" here (instead of simply passing the file names and count every non-option a config file), how to specify multiple config files (rare thing to be needed, but just in case – example use-case: "device A for the quarterly trip to Turkey", so first load device-specific file, then country-specific one)? Multiple `-C | --config` arguments processed in the order specified?

So if @n76 agrees I'd say go for 1-4. Let me know if you want my help on that (though I guess you're faster with this, as the current structure of "modules" is your work I'd need to familarize myself with first).

**Comment by n76** [comment]
Sorry I have been silent on this. . .

My thoughts:

1. Having a .gitignore to exclude all files not specifically desired in the repository is a very good thing.

2. Processing include(s) in the order 1 through 4 is reasonable. A default include should always be loaded first so that all items are defined and then later includes can override values as desired. I wouldn't personally be using multiple configurations for multiple devices but I can see where that might be useful. And I can see where setting up a special config for a trip with only the few items to be changed is useful (I could have used that for my trip to Europe last month).

If either of you wishes to make those changes, please do. In this case I'd hope for comments from all before I did the merge so that issues like the .gitignore will not reoccur.

**Comment by IzzySoft** [comment]
Full ack to that, @n76 !

@wvengen will you go for the "1-4" steps then? Shall I take care for the `.gitignore` meanwhile (at least the "general things" we've already discussed – like database files, tar archives, and CSV files (making sure to not affect those in the `misc/` folder, of course, but rather limit it to the ones downloaded/generated – which should not end up in `misc/`) – or better leave this to you as well?

**Comment by wvengen** [comment]
I'll have a look, great!

Regardig `.gitignore` - all generated files are already in there (unless I forgot something). I suppose you assume users are shuffling around csv files, and you would like to prevent them from accidentaly checking them in?

**Comment by IzzySoft** [comment]
@wvengen Nope. As I wrote, for "normal users" the `.gitignore` is irrelevant (they don't check in anything, neither incidentally nor accidentally). It's just important for "contributors" (us) who should know what they do. I'd cover the "default locations" (where those files are created if not explicitly specified otherwise), so no data from "test runs" can get accidentally added/committed.

**Comment by wvengen** [comment]
Ah, thanks. Sounds good!

I'm realizing something: when parameters are passed, environment variables are set within `lacells-creator` to help determining what needs to happen, but this is not a one-to-one mapping. A configuration file could set these variables as well, but it would be somewhat different from the command-line parameters. I think that's confusing. And it's an internal API that may change in the future.

We could just support the `API_KEY` (already) and `MCC` options and leave the rest, but that would be a bit odd for a configuration file and not extensible. Or provide an `ARGS` variable for command-line arguments (ugly). Or directly translate command-line arguments to internal environment variables that make more sense and allow those to be set from a configuration file too, then do all processing of the arguments (more refactoring than I'm happy with at this moment).

Not sure how to proceed now.

**Comment by IzzySoft** [comment]
I'm for the last approach. Summed up:

1. Parse CLI options and separate `-C | --config`
1. Read the config files in the order discussed above (including those specified at CLI)
1. Evaluate *remaining* command-line parameters which have corresponding variables and map them accordingly

This would still mean *parsing* the command-line parameters before reading config files as those might be passed that way – but as options passed by CLI should have precendence over values from config files, these other ones must be mapped *after* the config files have been "sourced".

I'm also fine with the `ARGS` variant. Important (from the user's point of view) is the *result* must be equivalent (e.g. if there's a `MERGE=1` setting, that should have the same effect as specifying the `-m` option at CLI).

Also keep in mind there are CLI options not having a corresponding config setting (at least I'm not sure if such a setting would make sense for `--export | --import`, i.e. for "action switches", but I'm very sure it makes no sense for `--help`). I also tend to skip settings for `--download`, though I'm undecided for `--push`: if we add settings for these two, wouldn't that mean we'd also need to have a corresponding `--no*` CLI option (i.e. `--no-download | --no-push`) to override them?

So basically, I'd start with the API_KEY, MCC, and (optionally) MNC settings (oops: no CLI switch for the latter?) – which should be pretty straight-forward.

Considering whether to add a CLI option for `-n | --filter-mnc` (completely missing according to `misc/usage.txt` – but I just realized we didn't have that before, it was a different script I've used in the past offering this; not crucial, but nice-to-have) as well as other additional options/settings (e.g. `-v | --verbose` (maybe "shrink" output from e.g. `wget` when not in "verbose mode"), `--no-quiet` (and other "invertors") *may* be a separate issue later, and dealt with then (when needed/requested).

**Comment by wvengen** [comment]
Thanks!
An mnc filter might be nice! Perhaps we could specify it as MCC:MNC[:LAC], preferably with ranges like 123:10-150. Something for a later enhancement.

I've kept it with the three documented options (`API_KEY`, `MCC` and `QUIET`) as mostly discussed here. Config file locations almost as proposed, if you have other ideas please let me know.

**Comment by IzzySoft** [comment]
Sounds phantastic, thanks a lot!

**Comment by n76** [comment]
Discussion moved to https://github.com/n76/lacells-creator/pull/14
